### DISCUSSION TITLE
How would you create a dependency key for something that has an async initializer?

### DISCUSSION THREAD

**Discussion by BrentMifsud** [pointfreeco/swift-dependencies#271] [Q&A] [answered]
For example, CLMonitor:

![image](https://github.com/user-attachments/assets/b2aed208-937b-4200-b574-351318a0bf3f)

**Comment by stephencelis** [comment, marked as answer]
Generally you should consider dependency values as static interfaces, so the key could be of a sendable closure instead:

```swift
static var liveValue: @Sendable () async -> CLMonitor = {
  await CLMonitor("my-location-monitor)
}
```

Or if you need a long-living monitor, you could wrap it in an actor:

```swift
actor Monitor {
  private var _rawValue: CLMonitor?

  var rawValue: CLMonitor {
    get async {
      guard let _rawValue else {
        let rawValue = await CLMonitor("my-location-monitor")
        _rawValue = rawValue
        return rawValue
      }
      return _rawValue
    }
  }
}

// ...
static var liveValue: Monitor { Monitor() }
```

More work would need to be done to make things testable, though. Check out [Designing dependencies](https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/designingdependencies) for more info.

**Reply by BjornRuud** [reply]
I have one value which needs async init but access to it does not require async. I created a key with a static var that can be set by the app init and resolve it in `liveValue`:

```swift
enum AppEnvironmentKey: DependencyKey {
    nonisolated(unsafe) static var _liveValue: Environment?

    static let liveValue: Environment = {
        guard let _liveValue else {
            fatalError("Environment not ready.")
        }
        return _liveValue
    }()
}
```

Then in app init before dependencies are used:

```swift
public func buildApplication(_ arguments: some AppArguments) async throws -> some ApplicationProtocol {
    switch DependencyValues._current.context {
    case .live:
        AppEnvironmentKey._liveValue = Environment().merging(with: try await .dotEnv())
    default:
        break
    }
}
```

How terrible is this? 😄 What options do we have for values that cannot be initialized directly in `liveValue` and lives for the lifetime of the app? Wrapping in an actor doesn't work for me because then the value has an async API.

**Reply by BrentMifsud** [reply]
As @stephencelis mentioned above, you can make the live value a closure.

```
var liveValue: () async -> MyDependency = { 
    await MyDependency() 
}
```

**Reply by BjornRuud** [reply]
Sure, but then I would have to await the use of the live value and I'm back to square one. My use case is async init but non-async access and usage.
